homeassistant/components/madvr/remote.py

```diff
@@ -0,0 +1,106 @@
+"""Support for MadVR remote control."""
+
+from collections.abc import Iterable
+import logging
+from typing import Any
+
+from madvr.madvr import Madvr
+
+from homeassistant.components.remote import RemoteEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+from .coordinator import MadVRCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the MadVR remote."""
+    coordinator: MadVRCoordinator = entry.runtime_data
+    if not isinstance(coordinator, MadVRCoordinator):
+        raise TypeError("entry.runtime_data is not an instance of MadVRCoordinator")
+
+    madvr_client = coordinator.client
+
+    async_add_entities(
+        [
+            MadvrRemote(hass, coordinator, madvr_client, entry.entry_id),
+        ]
+    )
+
+
+class MadvrRemote(CoordinatorEntity, RemoteEntity):
+    """Remote entity for the MadVR integration."""
+
+    coordinator: MadVRCoordinator
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        coordinator: MadVRCoordinator,
+        madvr_client: Madvr,
+        entry_id: str,
+    ) -> None:
+        """Initialize the remote entity."""
+        super().__init__(coordinator)
+        self.madvr_client = madvr_client
+        self.coordinator = coordinator
+        self._attr_name = coordinator.name
+        self._attr_unique_id = f"{entry_id}_remote"
+        self.entry_id = hass.data[DOMAIN]["entry_id"]
+        self._attr_should_poll = False
+        self.connection_event = self.madvr_client.connection_event
+        self.madvr_client.set_update_callback(coordinator.handle_push_data)
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        await super().async_added_to_hass()
+        # inform the client to start tasks
+        _LOGGER.debug("Adding to hass")
+        _LOGGER.debug("Using loop: %s", self.madvr_client.loop)
+        await self.madvr_client.async_add_tasks()
+        _LOGGER.debug("Added %s tasks to client", len(self.madvr_client.tasks))
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when removed."""
+        _LOGGER.debug("Removing from hass")
+        # inform the client to cancel all tasks
+        await self.coordinator.async_handle_unload()
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if the device is on."""
+        return self.madvr_client.is_on
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return extra state attributes."""
+        # Useful for making sensors
+        return self.madvr_client.msg_dict
```

Comment:
If they are useful for making sensors, let's make them proper sensors instead. We want to add as little unneeded extra state attributes as possible

Reply:
okay I can remove this. The original PR had sensors and binary sensors but the bot made me remove them